tortoise/tortoise.go

```diff
@@ -1108,151 +890,95 @@ func (t *turtle) ballotFilterForHealing(logger log.Log) func(types.BallotID) boo
 		if !exist {
 			logger.With().Error("inconsistent state: ballot not found", ballotID)
 			return false
 		}
-		return t.Last.Difference(lid) > t.BadBeaconVoteDelayLayers
-	}
-}
-
-func getEpochWeight(atxdb atxDataProvider, epochWeight map[types.EpochID]uint64, eid types.EpochID) (uint64, error) {
-	weight, exist := epochWeight[eid]
-	if exist {
-		return weight, nil
-	}
-	weight, _, err := atxdb.GetEpochWeight(eid)
-	if err != nil {
-		return 0, fmt.Errorf("epoch weight %s: %w", eid, err)
-	}
-	epochWeight[eid] = weight
-	return weight, nil
-}
-
-// computeExpectedVoteWeight computes the expected weight for a block in the target layer, from summing up votes from ballots up to and including last layer.
-func computeExpectedVoteWeight(atxdb atxDataProvider, epochWeight map[types.EpochID]uint64, target, last types.LayerID) (*big.Float, error) {
-	var (
-		total  = new(big.Float)
-		layers = uint64(types.GetLayersPerEpoch())
-	)
-	// TODO(dshulyak) this can be improved by computing weight for epoch at once
-	for lid := target.Add(1); !lid.After(last); lid = lid.Add(1) {
-		weight, err := getEpochWeight(atxdb, epochWeight, lid.GetEpoch())
-		if err != nil {
-			return nil, err
-		}
-		layerWeight := new(big.Float).SetUint64(weight)
-		layerWeight.Quo(layerWeight, new(big.Float).SetUint64(layers))
-		total.Add(total, layerWeight)
+		return t.last.Difference(lid) > t.BadBeaconVoteDelayLayers
 	}
-	return total, nil
 }
 
 // computeBallotWeight compute and assign ballot weight to the weights map.
 func computeBallotWeight(
 	atxdb atxDataProvider,
-	weights map[types.BallotID]*big.Float,
+	weights map[types.BallotID]weight,
 	ballot *types.Ballot,
 	layerSize,
 	layersPerEpoch uint32,
-) (*big.Float, error) {
+) (weight, error) {
 	if ballot.EpochData != nil {
-		var total, weight uint64
+		var total, targetWeight uint64
 
 		for _, atxid := range ballot.EpochData.ActiveSet {
 			atx, err := atxdb.GetAtxHeader(atxid)
 			if err != nil {
-				return nil, fmt.Errorf("atx %s in active set of %s is unknown", atxid, ballot.ID())
+				return weightFromUint64(0), fmt.Errorf("atx %s in active set of %s is unknown", atxid, ballot.ID())
 			}
 			atxweight := atx.GetWeight()
 			total += atxweight
 			if atxid == ballot.AtxID {
-				weight = atxweight
+				targetWeight = atxweight
 			}
 		}
 
-		expected, err := proposals.GetNumEligibleSlots(weight, total, layerSize, layersPerEpoch)
+		expected, err := proposals.GetNumEligibleSlots(targetWeight, total, layerSize, layersPerEpoch)
 		if err != nil {
-			return nil, fmt.Errorf("unable to compute number of eligibile ballots for atx %s", ballot.AtxID)
+			return weightFromUint64(0), fmt.Errorf("unable to compute number of eligibile ballots for atx %s", ballot.AtxID)
 		}
-		rst := new(big.Float).SetUint64(weight)
-		rst.Quo(rst, new(big.Float).SetUint64(uint64(expected)))
-		return rst, nil
+		rst := weightFromUint64(targetWeight)
+		return rst.div(weightFromUint64(uint64(expected))), nil
```

Comment:
it is handled in GetNumEligibleSlots, so that if expected is 0 and total weight is not 0 - it is always at least 1 ballot. i guess i will just error for safety